library/std/src/sys/uefi/stdio.rs

```diff
@@ -4,45 +4,48 @@ use crate::mem::MaybeUninit;
 use crate::os::uefi;
 use crate::ptr::NonNull;
 
-const MAX_BUFFER_SIZE: usize = 8192;
+pub struct Stdin {
+    pending: Option<char>,
+}
 
-pub struct Stdin;
 pub struct Stdout;
 pub struct Stderr;
 
 impl Stdin {
     pub const fn new() -> Stdin {
-        Stdin
+        Stdin { pending: None }
     }
 }
 
 impl io::Read for Stdin {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+    fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
         let st: NonNull<r_efi::efi::SystemTable> = uefi::env::system_table().cast();
         let stdin = unsafe { (*st.as_ptr()).con_in };
 
-        // Try reading any pending data
-        let inp = match read_key_stroke(stdin) {
-            Ok(x) => x,
-            Err(e) if e == r_efi::efi::Status::NOT_READY => {
-                // Wait for keypress for new data
-                wait_stdin(stdin)?;
-                read_key_stroke(stdin).map_err(|x| io::Error::from_raw_os_error(x.as_usize()))?
+        // Write any pending character
+        if let Some(ch) = self.pending {
+            if ch.len_utf8() > buf.len() {
+                return Ok(0);
             }
-            Err(e) => {
-                return Err(io::Error::from_raw_os_error(e.as_usize()));
-            }
-        };
+            ch.encode_utf8(buf);
+            buf = &mut buf[ch.len_utf8()..];
+            self.pending = None;
+        }
+
+        // Try reading any pending data
+        let inp = simple_text_input_read(stdin)?;
 
         // Check if the key is printiable character
-        if inp.scan_code != 0x00 {
+        if inp == 0x00 {
             return Err(io::const_io_error!(io::ErrorKind::Interrupted, "Special Key Press"));
```

Comment:
I think you want to suppress these. No other implementation returns errors if `ctrl` or `capslock` are pressed, so neither should UEFI. This is not a keyboard interface, so I think the correct call is to suppress unknown scancodes.

Reply:
Isn't that because others have an actual terminal that works out how to respond to those cases?

Reply:
We can maybe phrase it differently: What is a user of `Stdin` supposed to do when getting this error code, other than ignore it? The actual data has already been discarded, there is no way to recover it. The error code carries no real information. I don't see the purpose of returning this error.

Reply:
I do agree that, under most circumstances, it makes sense to ignore the error. However, I have 2 cases that make me reconsider:
1. UEFI shell does not provide any feedback when reading from stdin (so no indication of what you are typing). This means that if we suppress the non-printable characters, the user will have no way of knowing (unless they know the internals). It might be especially problematic in case of a faulty keyboard.
2. Piping the input of an invalid file might lead to everything being suppressed instead of any error.

Reply:
I think we could defer this question to a later PR if settling it is going to hold up fixing the byte stream handling.